src/index/PatternIndex.cpp

```diff
@@ -0,0 +1,834 @@
+#include "PatternIndex.h"
+#include "MetaDataIterator.h"
+
+const uint32_t PatternIndex::PATTERNS_FILE_VERSION = 1;
+const uint32_t PatternIndex::NAMESPACE_FILE_VERSION = 0;
+
+const std::string PatternIndex::PATTERNS_NAMESPACE_SUFFIX =
+    ".index.patterns.pred_namespace";
+const std::string PatternIndex::PATTERNS_SUBJECTS_SUFFIX = ".index.patterns";
+const std::string PatternIndex::PATTERNS_OBJECTS_SUFFIX =
+    ".index.patterns.objects";
+
+// TODO: stxxl::vector<array<Id, 3>> should be Index::TripleVec but that
+// creates a circular dependency
+using TripleVec = stxxl::vector<array<Id, 3>>;
+
+PatternIndex::PatternIndex()
+    : _maxNumPatterns(std::numeric_limits<PatternID>::max() - 2),
+      _num_bytes_per_predicate(0),
+      _initialized(false) {}
+
+// _____________________________________________________________________________
+void PatternIndex::generatePredicateLocalNamespace(
+    VocabularyData* vocabData, const std::string& filename_base) {
+  // Create a new namespace for predicates containing only predicates.
+  // This will be significantly smaller than the global namespace which
+  // also contains subjects and objects, and allows for shrinking
+  // the pattern trick data.
+
+  // This is not inside a templated method as the PSO metadata is based upon
+  // HashMaps which need to be treated differently
+  TripleVec::bufreader_type reader(*vocabData->idTriples);
```

Comment:
I don't understand this comment, and I don't think it is necessary. We don't deal with the PSO metadata here at all.